File: sdks/go/sync/sync.go
package sync

import (
	"context"
	"fmt"
	"time"

	"github.com/Colin4k1024/Oris/sdks/go/experience"
	"github.com/Colin4k1024/Oris/sdks/go/hub"
	"github.com/Colin4k1024/Oris/sdks/go/store"
)

type Config struct {
	Store      *store.LocalStore
	Experience *experience.Client
	Hub        *hub.Client
}

type SyncManager struct {
	store      *store.LocalStore
	experience *experience.Client
	hub        *hub.Client
}

func New(cfg Config) *SyncManager {
	return &SyncManager{
		store:      cfg.Store,
		experience: cfg.Experience,
		hub:        cfg.Hub,
	}
}

func (m *SyncManager) PushToHub(ctx context.Context, opts store.PushOpts) (*store.PushResult, error) {
	if m.experience == nil {
		return nil, fmt.Errorf("experience client not configured")
	}

	var genes []store.Gene
	var err error

	if len(opts.GeneIDs) > 0 {
		for _, id := range opts.GeneIDs {
			g, err := m.store.Get(ctx, id)
			if err != nil {
				return nil, fmt.Errorf("get gene %s: %w", id, err)
			}
			if g != nil {
				genes = append(genes, *g)
			}
		}
	} else {
		genes, err = m.store.GetUnsynced(ctx)
		if err != nil {
			return nil, fmt.Errorf("get unsynced: %w", err)
		}
	}

	result := &store.PushResult{}
	for _, g := range genes {
		payload := geneToPayload(g)
		_, shareErr := m.experience.Share(ctx, payload)

		entry := store.SyncLogEntry{
			Direction: "push",
			GeneID:    g.GeneID,
			Timestamp: time.Now().UTC(),
		}

		if shareErr != nil {
			entry.Status = "failed"
			entry.ErrorMessage = shareErr.Error()
			result.Failed++
			result.Errors = append(result.Errors, store.PushError{
				GeneID:  g.GeneID,
				Message: shareErr.Error(),
			})
		} else {
			entry.Status = "success"
			result.Pushed++
			m.store.MarkSynced(ctx, g.GeneID, time.Now().UTC())
		}
		m.store.LogSync(ctx, entry)
	}
	return result, nil
}

func (m *SyncManager) PullFromHub(ctx context.Context, opts store.PullOpts) (int, error) {
	if m.experience == nil {
		return 0, fmt.Errorf("experience client not configured")
	}

	query := &experience.FetchQuery{
		Q:             opts.Q,
		MinConfidence: opts.MinConfidence,
		Limit:         opts.Limit,
	}

	resp, err := m.experience.Fetch(ctx, query)
	if err != nil {
		return 0, fmt.Errorf("fetch from hub: %w", err)
	}

	imported := 0
	for _, asset := range resp.Assets {
		gene := assetToGene(asset)

		existing, err := m.store.Get(ctx, gene.GeneID)
		if err != nil {
			continue
		}

		if existing != nil {
			if coreFieldsMatch(existing, &gene) {
				gene = mergeStats(*existing, gene)
			} else {
				m.store.LogSync(ctx, store.SyncLogEntry{
					Direction:    "pull",
					GeneID:       gene.GeneID,
					Status:       "conflict",
					ErrorMessage: "core fields differ, skipped",
					Timestamp:    time.Now().UTC(),
				})
				continue
			}
		}

		if err := m.store.Save(ctx, gene); err != nil {
			continue
		}

		m.store.LogSync(ctx, store.SyncLogEntry{
			Direction: "pull",
			GeneID:    gene.GeneID,
			Status:    "success",
			Timestamp: time.Now().UTC(),
		})
		imported++
	}
	return imported, nil
}

func (m *SyncManager) RegisterNode(ctx context.Context, req *hub.RegisterRequest) (*hub.RegisterResponse, error) {
	if m.hub == nil {
		return nil, fmt.Errorf("hub client not configured")
	}
	return m.hub.Register(ctx, req)
}

func (m *SyncManager) GetSyncLog(ctx context.Context, limit int) ([]store.SyncLogEntry, error) {
	return m.store.GetSyncLog(ctx, limit)
}

func geneToPayload(g store.Gene) map[string]any {
	p := map[string]any{
		"type":          "gene",
		"id":            g.GeneID,
		"confidence":    g.Confidence,
		"quality_score": g.QualityScore,
		"use_count":     g.UseCount,
		"success_count": g.SuccessCount,
		"created_at":    g.CreatedAt.Format(time.RFC3339),
	}
	if g.Strategy != nil {
		p["strategy"] = g.Strategy
	}
	if g.Signals != nil {
		p["signals"] = g.Signals
	}
	if g.Validation != nil {
		p["validation"] = g.Validation
	}
	if g.ContributorID != "" {
		p["contributor_id"] = g.ContributorID
	}
	return p
}

func assetToGene(a experience.NetworkAsset) store.Gene {
	now := time.Now().UTC()
	g := store.Gene{
		GeneID:        a.ID,
		Name:          a.ID,
		TaskClass:     "unknown",
		Confidence:    a.Confidence,
		QualityScore:  a.QualityScore,
		UseCount:      a.UseCount,
		SuccessCount:  a.SuccessCount,
		ContributorID: a.ContributorID,
		Source:        "hub",
		CreatedAt:     a.CreatedAt,
		UpdatedAt:     now,
		SyncedAt:      &now,
	}

	if s, ok := a.Strategy.(map[string]any); ok {
		g.Strategy = s
	}
	if s, ok := a.Signals.(map[string]any); ok {
		g.Signals = s
	}
	if s, ok := a.Validation.(map[string]any); ok {
		g.Validation = s
	}

	return g
}

func coreFieldsMatch(existing, incoming *store.Gene) bool {
	return existing.TaskClass == incoming.TaskClass
}

func mergeStats(existing, incoming store.Gene) store.Gene {
	if incoming.UseCount > existing.UseCount {
		existing.UseCount = incoming.UseCount
	}
	if incoming.SuccessCount > existing.SuccessCount {
		existing.SuccessCount = incoming.SuccessCount
	}
	if incoming.Confidence > existing.Confidence {
		existing.Confidence = incoming.Confidence
	}
	if incoming.QualityScore > existing.QualityScore {
		existing.QualityScore = incoming.QualityScore
	}
	existing.UpdatedAt = time.Now().UTC()
	now := time.Now().UTC()
	existing.SyncedAt = &now
	return existing
}

File: sdks/typescript/src/sync.ts
import type { ExperienceClient } from "./experience.js";
import type { HubClient } from "./hub.js";
import type { Gene, PullOpts, PushError, PushOpts, PushResult, SyncLogEntry } from "./gene.js";
import type { LocalStore } from "./store.js";
import type { NetworkAsset } from "./types.js";

export interface SyncConfig {
  store: LocalStore;
  experience?: ExperienceClient;
  hub?: HubClient;
}

export class SyncManager {
  private store: LocalStore;
  private experience?: ExperienceClient;
  private hub?: HubClient;

  constructor(config: SyncConfig) {
    this.store = config.store;
    this.experience = config.experience;
    this.hub = config.hub;
  }

  async pushToHub(opts?: PushOpts): Promise<PushResult> {
    if (!this.experience) throw new Error("experience client not configured");

    let genes: Gene[];
    if (opts?.geneIds && opts.geneIds.length > 0) {
      genes = opts.geneIds
        .map((id) => this.store.get(id))
        .filter((g): g is Gene => g !== null);
    } else {
      genes = this.store.getUnsynced();
    }

    const result: PushResult = { pushed: 0, failed: 0, errors: [] };

    for (const gene of genes) {
      const payload = geneToPayload(gene);
      const entry: Omit<SyncLogEntry, "id"> = {
        direction: "push",
        geneId: gene.geneId,
        status: "",
        remoteUrl: "",
        errorMessage: "",
        timestamp: new Date(),
      };

      try {
        await this.experience.share(payload);
        entry.status = "success";
        result.pushed++;
        this.store.markSynced(gene.geneId, new Date());
      } catch (e) {
        const msg = e instanceof Error ? e.message : String(e);
        entry.status = "failed";
        entry.errorMessage = msg;
        result.failed++;
        result.errors.push({ geneId: gene.geneId, message: msg });
      }

      this.store.logSync(entry);
    }
    return result;
  }

  async pullFromHub(opts?: PullOpts): Promise<number> {
    if (!this.experience) throw new Error("experience client not configured");

    const resp = await this.experience.fetch_({
      q: opts?.q,
      minConfidence: opts?.minConfidence,
      limit: opts?.limit,
    });

    let imported = 0;
    for (const asset of resp.assets) {
      let gene = assetToGene(asset);
      const existing = this.store.get(gene.geneId);

      if (existing) {
        if (coreFieldsMatch(existing, gene)) {
          gene = mergeStats(existing, gene);
        } else {
          this.store.logSync({
            direction: "pull",
            geneId: gene.geneId,
            status: "conflict",
            remoteUrl: "",
            errorMessage: "core fields differ, skipped",
            timestamp: new Date(),
          });
          continue;
        }
      }

      this.store.save(gene);
      this.store.logSync({
        direction: "pull",
        geneId: gene.geneId,
        status: "success",
        remoteUrl: "",
        errorMessage: "",
        timestamp: new Date(),
      });
      imported++;
    }
    return imported;
  }

  getSyncLog(limit = 50): SyncLogEntry[] {
    return this.store.getSyncLog(limit);
  }
}

function geneToPayload(g: Gene): Record<string, unknown> {
  const p: Record<string, unknown> = {
    type: "gene",
    id: g.geneId,
    confidence: g.confidence,
    quality_score: g.qualityScore,
    use_count: g.useCount,
    success_count: g.successCount,
    created_at: g.createdAt.toISOString(),
  };
  if (g.strategy) p.strategy = g.strategy;
  if (g.signals) p.signals = g.signals;
  if (g.validation) p.validation = g.validation;
  if (g.contributorId) p.contributor_id = g.contributorId;
  return p;
}

function assetToGene(a: NetworkAsset): Gene {
  const now = new Date();
  return {
    geneId: a.id,
    name: a.id,
    taskClass: "unknown",
    confidence: a.confidence ?? 0,
    qualityScore: a.quality_score ?? 0,
    useCount: a.use_count ?? 0,
    successCount: a.success_count ?? 0,
    contributorId: a.contributor_id ?? "",
    source: "hub",
    syncedAt: now,
    createdAt: a.created_at ? new Date(a.created_at) : now,
    updatedAt: now,
  };
}

function coreFieldsMatch(existing: Gene, incoming: Gene): boolean {
  return existing.taskClass === incoming.taskClass;
}

function mergeStats(existing: Gene, incoming: Gene): Gene {
  return {
    ...existing,
    useCount: Math.max(existing.useCount, incoming.useCount),
    successCount: Math.max(existing.successCount, incoming.successCount),
    confidence: Math.max(existing.confidence, incoming.confidence),
    qualityScore: Math.max(existing.qualityScore, incoming.qualityScore),
    updatedAt: new Date(),
    syncedAt: new Date(),
  };
}
